packages/typespec-ts/src/utils/modelUtils.ts

```diff
@@ -590,10 +599,12 @@ function getSchemaForModel(
   }
   modelSchema.properties = {};
 
-  const derivedModels = model.derivedModels.filter(includeDerivedModel);
-
   // getSchemaOrRef on all children to push them into components.schemas
   const discriminator = getDiscriminator(program, model);
+  // should respect needRef for derived models unless there's a discriminator in base model
+  const derivedModels = model.derivedModels.filter((dm) => {
+    return includeDerivedModel(dm, discriminator ? false : needRef);
+  });
```

Comment:
this is the reason why we get maxium callback stack error, because previously, we will call error's derived models, and then in it's derived models processing, we will handle base model error again.

Reply:
Could we add a UT to cover this?